src/wallet/spend.cpp

```diff
@@ -799,6 +815,83 @@ util::Result<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& av
     return op_selection_result;
 }
 
+/** Returns UTXO targets and compute their current count from available coins */
+std::vector<UtxoTarget> UtxoTargetsFromJson(const UniValue& utxo_targets_json, const CoinsResult& available_coins)
+{
+    std::vector<UtxoTarget> utxo_targets;
+    for (const auto& bucket : utxo_targets_json.getValues()) {
+        UtxoTarget target = {
+            bucket["start_satoshis"].getInt<CAmount>(),
+            bucket["end_satoshis"].getInt<CAmount>(),
+            bucket["target_utxo_count"].getInt<uint32_t>()};
+        utxo_targets.push_back(target);
+    }
+    for (const auto& [type, outputs] : available_coins.coins) {
+        for (const wallet::COutput& output : outputs) {
+            for (UtxoTarget& target : utxo_targets) {
+                if (output.txout.nValue >= target.start_satoshis && output.txout.nValue <= target.end_satoshis) {
+                    target.current_utxo_count++;
+                    break;
+                }
+            }
+        }
+    }
+
+    return utxo_targets;
+}
+
+/** Returns a random change amount in the range of the most depleted Utxo bucket and sets `capacity`
+ * to the capacity of that change target, if any.
+ */
+std::optional<CAmount> GenerateChangeTargetFromUtxoTargets(const std::vector<UtxoTarget>& utxo_targets, const CAmount change_fee, double& capacity, FastRandomContext& rng)
+{
+    // Refill the most depleted utxo targets first.
+    auto target_iter = std::min_element(utxo_targets.begin(), utxo_targets.end(), [](const UtxoTarget& a, const UtxoTarget& b) {
+        return (double(a.current_utxo_count) / a.target_utxo_count) < (double(b.current_utxo_count) / b.target_utxo_count);
+    });
+    std::optional<CAmount> change_target;
+    if (target_iter != utxo_targets.end()) {
+        // Adjust the range of the target bucket based on the current fee rate.
+        change_target = rng.randrange(target_iter->end_satoshis - target_iter->start_satoshis - change_fee) + target_iter->start_satoshis + change_fee;
+        capacity = double(target_iter->current_utxo_count) / target_iter->target_utxo_count;
+    }
+    return change_target;
+}
+
+std::list<CTxOut> SplitChangeFromUtxoTargets(CAmount change_amount, std::vector<UtxoTarget> utxo_targets, CAmount change_target, const CAmount change_fee, FastRandomContext& rng, CScript script_change)
```

Comment:
I'm wondering whether we should use a fully deterministic algorithm to split the change into buckets, I'm afraid this is exactly the kind of algorithm that may end up in a loop where one funding attempt does something, and the next one undoes it. It feels like adding randomization may be the simple way to avoid this? But maybe it's too early in the algorithm design phase to decide.

Reply:
I agree that there's a risk that one funding operation adds an extra input from an existing bucket and then creates a change output to refill that same bucket, for example. My intuition though is that we can't optimize too far into the future.

Even if an input is cannibalized from a bucket, it will be "reforged" into a new utxo that hopefully has a value that is more likely to be selected by BnB as a single input to a future changeless funding transaction.  The current scheme randomizes new change outputs and skews them based on the current fee rate to try to optimize for future changeless single input transactions.

Can we do better with some ideas from memory caching? Maybe use the frequency that a particular amount is requested for funding to influence whether it is used as an extra input or as the size of a new change output?